core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java

```diff
@@ -0,0 +1,321 @@
+      } else {
+        return either.getRight();
+      }
+    };
+  }
+
+  private BiConsumer<Throwable, Object> getRetryHandler() {
+    return (error, offendingEvent) -> {
+      MessagingException messagingError = (MessagingException) error;
+      RetryContext ctx = popRetryContext(messagingError.getEvent());
+      int retriesLeft =
+          ctx.retryCount.getAndDecrement();
+      if (retriesLeft > 0) {
+        LOGGER.error("Retrying execution of event, attempt {} of {}.", ctx.getRetriesLeft(),
+                     ctx.maxRetries != RETRY_COUNT_FOREVER ? ctx.maxRetries : "unlimited");
+
+        // Schedule retry with delay
+        UntilSuccessfulRouter.this.delayScheduler.schedule(() -> innerRecorder.next(pushRetryContext(ctx.event, ctx)),
+                                                           ctx.delayInMillis, TimeUnit.MILLISECONDS);
+      } else { // Retries exhausted
+        // Current context already pooped. No need to re-insert it
+        LOGGER.error("Retry attempts exhausted. Failing...");
+        Throwable resolvedError = getThrowableFunction(ctx.event).apply(error);
+        downstreamRecorder.next(Either.left(resolvedError, CoreEvent.class));
+        completeRouterIfNecessary();
+      }
+    };
+  }
+
+  /**
+   * If there are no events in-flight and the upstream publisher has received a completion signal, complete downstream publishers.
+   */
+  private void completeRouterIfNecessary() {
+    if (completeDeferred.get() && inflightEvents.get() == 0) {
+      completeRouter();
+    }
+  }
+
+  /**
+   * Complete both downstream publishers.
+   */
+  private void completeRouter() {
+    innerRecorder.complete();
+    downstreamRecorder.complete();
+  }
+
+  /**
+   * Assembles and returns the downstream {@link Publisher<CoreEvent>}.
+   * 
+   * @return the successful {@link CoreEvent} or retries exhaustion errors {@link Publisher}
+   */
+  Publisher<CoreEvent> getDownstreamPublisher() {
+    return downstreamFlux
+        .compose(downstreamPublisher -> Mono.subscriberContext()
+            .flatMapMany(downstreamContext -> downstreamPublisher.doOnSubscribe(s -> {
+              innerFlux.subscriberContext(downstreamContext).subscribe();
+              upstreamFlux.subscriberContext(downstreamContext).subscribe();
+            })));
+  }
+
+
+  /**
+   * Pushes the {@link RetryContext} inside the event being routed through the retrial chain.
+   * 
+   * @param event the current retrial {@link CoreEvent}
+   * @param ctx the current {@link RetryContext}
+   * @return the {@link CoreEvent} with the retry context saved as internal parameter
+   */
+  private CoreEvent pushRetryContext(CoreEvent event, RetryContext ctx) {
+    // Requires: The ctx that corresponds to this router execution is not in the stack, or there's no stack yet
+    // Assures: The ctx that corresponds to this router execution is the one on top of the stack
+
+    // TODO: There must be cleaner way to do this
```

Comment:
i has a similar issue for handling nested policies, we should extract that and reuse here